pkg/splunk/enterprise/util.go

```diff
@@ -34,20 +37,248 @@ import (
 // kubernetes logger used by splunk.enterprise package
 var log = logf.Log.WithName("splunk.enterprise")
 
-// ApplyCommonSecretObject creates/updates the namespace scoped "splunk-secrets" K8S secret object
-func ApplyCommonSecretObject(client splcommon.ControllerClient, namespace string) (*corev1.Secret, error) {
+// GetVersionedSecretName returns a versioned secret name
+func GetVersionedSecretName(versionedSecretIdentifier string, version string) string {
+	return fmt.Sprintf("%s-secret-v%s", versionedSecretIdentifier, version)
+}
+
+// GetNamespaceScopedSecretName gets namespace scoped secret name
+func GetNamespaceScopedSecretName(namespace string) string {
+	return fmt.Sprintf(namespaceScopedSecretName, namespace)
+}
+
+// GetNamespaceScopedSecret retreives namespace scoped secret
+func GetNamespaceScopedSecret(c splcommon.ControllerClient, namespace string) (*corev1.Secret, error) {
+	var namespaceScopedSecret corev1.Secret
+
+	// Check if a namespace scoped secret exists
+	namespacedName := types.NamespacedName{Namespace: namespace, Name: GetNamespaceScopedSecretName(namespace)}
+	err := c.Get(context.TODO(), namespacedName, &namespaceScopedSecret)
+	if err != nil {
+		// Didn't find it
+		return nil, err
+	}
+
+	return &namespaceScopedSecret, nil
+}
+
+// GetVersionedSecretVersion checks if the secretName includes the versionedSecretIdentifier and if so, extracts the version
+func GetVersionedSecretVersion(secretName string, versionedSecretIdentifier string) (int, error) {
+	// Extracting version from secret's name
+	version := strings.TrimPrefix(secretName, GetVersionedSecretName(versionedSecretIdentifier, ""))
+
+	// Check if the secretName includes the versionedSecretIdentifier
+	if version != secretName {
+		// Includes, version extracted, check if version number is valid
+		versionInt, err := strconv.Atoi(version)
+		if err != nil {
+			return -1, errors.New(nonIntegerVersionError)
+		}
+
+		// Versions should be > 0
+		if versionInt <= 0 {
+			return -1, errors.New(lessThanOrEqualToZeroVersionError)
+		}
+
+		return versionInt, nil
+	}
+
+	// Secret name not matching required criteria
+	return -1, fmt.Errorf(nonMatchingStringError, secretName, versionedSecretIdentifier)
+}
+
+// GetExistingLatestVersionedSecret retreives latest EXISTING versionedSecretIdentifier based secret existing currently in the namespace
+func GetExistingLatestVersionedSecret(c splcommon.ControllerClient, namespace string, versionedSecretIdentifier string) (*corev1.Secret, int, bool) {
+	scopedLog := log.WithName("GetExistingLatestVersionedSecret").WithValues(
+		"versionedSecretIdentifier", versionedSecretIdentifier,
+		"namespace", namespace)
+
+	// Get list of secrets in K8S cluster
+	secretList := corev1.SecretList{}
+
+	// Retreive only secrets only from namespace
+	listOpts := []client.ListOption{
+		client.InNamespace(namespace),
```

Comment:
Just a thought is I wonder if any more criteria could be safely added to this query filter, such as owner is splunk operator?

Reply:
As per our discussion, list options doesn't have such a utility [reflink](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/client#ListOption).